In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Problem 5: Document Clustering

Document clustering is an important application of clustering algorithms.
Document clustering algorithms can be used to automatically group search results into categories, making it easier to find relevant results.

In this problem, you'll be clustering articles published on [Daily Kos](https://www.dailykos.com/), an American political blog

The file *dailykos.csv* contains data on 3,430 news articles or blogs that have been posted on Daily Kos.
These articles were posted in 2004, leading up to the United States Presidential Election. 

In [2]:
# load the data
url = 'https://raw.githubusercontent.com/um-perez-alvaro/Data-Science-Theory/master/Data/dailykos.csv'
data = pd.read_csv(url)
data

abandon  abc  ability  abortion  absolute  abstain  abu  abuse  accept  \
0           0    0        0         0         0        0    0      0       0   
1           0    0        0         0         0        0    0      0       0   
2           0    0        0         0         0        1    0      0       0   
3           0    0        0         0         0        0    0      0       0   
4           0    0        0         0         0        0    0      0       0   
...       ...  ...      ...       ...       ...      ...  ...    ...     ...   
3425        0    0        0         0         0        0    0      0       0   
3426        0    0        0         0         0        0    0      0       0   
3427        0    0        0         0         0        0    0      0       0   
3428        0    0        0         0         0        0    0      0       0   
3429        0    0        0         0         0        0    0      0       0   

      access  ...  yeah  year  yesterday  york  youll  young  youre  youve  \
0          0  ...     0     0          0     0      0      0      0      0   
1          0  ...     0     0          0     0      0      0      0      0   
2          0  ...     0     0          0     0      0      0      0      0   
3          0  ...     0     0          2     0      0      1      0      0   
4          0  ...     0     1          1     0      0      1      0      0   
...      ...  ...   ...   ...        ...   ...    ...    ...    ...    ...   
3425       0  ...     0     0          0     0      0      0      0      0   
3426       0  ...     0     0          0     0      0      0      0      0   
3427       0  ...     0     0          0     1      0      0      0      0   
3428       0  ...     0     0          0     0      0      0      0      0   
3429       0  ...     0     0          0     0      0      0      0      0   

      zogby  zone  
0         0     1  
1         0     0  
2         0     0  
3         0     0  
4         0     0  
...     ...   ...  
3425      0     0  
3426      0     0  
3427      0     0  
3428      0     0  
3429      0     0  

[3430 rows x 1545 columns]

Each of the features in the dataset is a word that has appeared in at least 50 different articles (1,545 words in total).  For each document, the feature values are the number of times that word appeared in the document.

In [4]:
list_of_words = np.array(data.columns)
list(list_of_words)

['abandon',
 'abc',
 'ability',
 'abortion',
 'absolute',
 'abstain',
 'abu',
 'abuse',
 'accept',
 'access',
 'accomplish',
 'account',
 'accurate',
 'accusations',
 'achieve',
 'acknowledge',
 'act',
 'action',
 'active',
 'activist',
 'actual',
 'add',
 'added',
 'addition',
 'address',
 'admin',
 'administration',
 'admit',
 'advance',
 'advantage',
 'advertise',
 'advised',
 'affair',
 'affect',
 'affiliate',
 'afghanistan',
 'afraid',
 'afternoon',
 'age',
 'agencies',
 'agenda',
 'agree',
 'ahead',
 'aid',
 'aim',
 'air',
 'alaska',
 'allegation',
 'allegory',
 'allied',
 'allowed',
 'alternative',
 'altsite',
 'amazing',
 'amendment',
 'america',
 'american',
 'amount',
 'amp',
 'analysis',
 'analyst',
 'anecdotal',
 'anger',
 'angry',
 'announce',
 'annual',
 'answer',
 'apologies',
 'apparent',
 'appeal',
 'appearance',
 'applied',
 'appointed',
 'approach',
 'approval',
 'apr',
 'april',
 'arab',
 'area',
 'arent',
 'arg',
 'argue',
 'argument',
 'arizona',
 'arm',
 'armstro

In this part, we are trying to cluster news articles or blog posts into groups. 
This can be used to show readers categories to choose from when trying to decide what to read. 

## Part 1: k-means clustering

Cluster the articles using k-means, and ensure that you have a good number of clusters (using the "elbow" method, for example).

In [17]:
X = data.to_numpy().astype('float')

## Part 2

 Find the 10 most frequent words in each cluster.

In 2004, one of the candidates for the Democratic nomination for the President of the United States was Howard Dean, John Kerry was the candidate who won the democratic nomination, and John Edwards with the running mate of John Kerry (the Vice President nominee). Given this information, which cluster best corresponds to the democratic party?

Which k-means cluster best corresponds to the Iraq War?

## Part 3: Hierarchical Clustering

Now, run hierarchical clustering (this will probably take you a while. Why?)
Pick the number of clusters equal to the number of clusters you used in parts 1 and 2.